File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.js
function solution(park, routes) {
  let startPoint;
  // 시작 위치 찾기
  for (let i = 0; i < park.length; i++) {
    for (let j = 0; j < park[0].length; j++) {
      if (park[i][j] === "S") {
        startPoint = [i, j];
      }
    }
  }

  const op = {
    E: [0, 1],
    W: [0, -1],
    S: [1, 0],
    N: [-1, 0],
  };

  routes.forEach((item, idx) => {
    const [pos, range] = item.split(" ");
    let currentPos = [...startPoint];
    let isChecked = true;

    for (let i = 0; i < range; i++) {
      currentPos[0] += op[pos][0];
      currentPos[1] += op[pos][1];

      // 장애물 판별
      if (
        currentPos[0] > park.length - 1 ||
        currentPos[0] < 0 ||
        currentPos[1] > park[0].length - 1 ||
        currentPos[1] < 0
      ) {
        isChecked = false;
        break;
      }

      if (park[currentPos[0]][currentPos[1]] === "X") {
        isChecked = false;
        break;
      }
    }
    if (isChecked) startPoint = currentPos;
  });

  return startPoint;
}

File: 프로그래머스/unrated/172928. 공원 산책/공원 산책.py
def getDirection(string):
    if string == "E":
        return 0
    elif string == "W":
        return 1
    elif string == "S":
        return 2
    else:
        return 3

def solution(park, routes):
    park = [list(i) for i in park]
    answer = []
#     # 동서남북
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]
    
    startX, startY = 0, 0
    for i in range(len(park)):
        for j in range(len(park[0])):
            if park[i][j] == "S":
                startX, startY = i, j
                
    for route in routes:
        direction, distance = route.split(" ")
        direction = getDirection(direction)
        distance = int(distance)
        
        tmpX, tmpY = startX, startY
        flag = False
        while distance:
            ntmpX, ntmpY = tmpX + dx[direction], tmpY + dy[direction]
            if 0 <= ntmpX < len(park) and 0 <= ntmpY < len(park[0]):
                if park[ntmpX][ntmpY] == "O":
                    tmpX, tmpY = ntmpX, ntmpY
                    distance -= 1
                elif park[ntmpX][ntmpY] == "X":
                    flag = True
                    break
            else:
                flag = True
                break
        
        if not flag:
            park[startX][startY] = "O"
            startX, startY = tmpX, tmpY
            
    return [startX, startY]